hazelcast/src/test/java/com/hazelcast/cache/TenantUnavailableTest.java

```diff
@@ -0,0 +1,181 @@
+                .createCache(cacheName, cacheConfig);
+        cache1.put(new KeyType(), new ValueType());
+        assertInstanceOf(ValueType.class, cache1.get(new KeyType()));
+
+        HazelcastInstance hz2 = factory.newHazelcastInstance(newConfig());
+        ICacheService cacheService = getCacheService(hz2);
+        disallowClassNames.add(KeyType.class.getName());
+        hz1.shutdown(); // force migration
+        CacheManager cacheManager = createServerCachingProvider(hz2).getCacheManager();
+        Cache<KeyType, ValueType> cache2 = cacheManager.getCache(cacheName);
+        disallowClassNames.clear();
+        assertInstanceOf(ValueType.class, cache2.get(new KeyType()));
+
+        destroyEventContext.get().tenantUnavailable();
+        disallowClassNames.add(KeyType.class.getName());
+
+        cacheConfig = cacheService.getCacheConfig(CACHE_MANAGER_PREFIX + cacheName);
+        Cache<KeyType, ValueType> cache3 = cacheManager.getCache(cacheName);
+        assertInstanceOf(ValueType.class, cache3.get(new KeyType()));
+        Assert.assertFalse("Class Loading Failed", classLoadingFailed);
+    }
+
+    @Test
+    public void testMigrationWithUnavailableClasses() throws InterruptedException {
+        classesAlwaysAvailable = false;
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(2);
+        HazelcastInstance hz1 = factory.newHazelcastInstance(newConfig());
+        CacheConfig<String, ValueType> cacheConfig = new CacheConfig<>();
+        cacheConfig.setTypes(String.class, ValueType.class);
+
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        CacheProxy<String, ValueType> cache1 = (CacheProxy) createServerCachingProvider(hz1)
+                .getCacheManager().createCache(cacheName, cacheConfig);
+        ValueType value = new ValueType();
+        putValuesToPartition(hz1, cache1, value, 0, 1);
+        putValuesToPartition(hz1, cache1, value, 1, 1);
+
+        disallowClassNames.add(ValueType.class.getName());
+        HazelcastInstance hz2 = factory.newHazelcastInstance(newConfig().setLiteMember(true));
+        hz1.getPartitionService().addMigrationListener(new MigrationListenerImpl());
+        CacheManager cacheManager = createServerCachingProvider(hz2).getCacheManager();
+        Cache<String, ValueType> cache2 = cacheManager.getCache(cacheName);
+        // force migration
+        hz2.getCluster().promoteLocalLiteMember();
+        latch.await(); // await migration
+        disallowClassNames.clear();
+        Iterator<Cache.Entry<String, ValueType>> it2 = cache2.iterator();
+
+        assertTrue("Iterator should not be empty", it2.hasNext());
+        while (it2.hasNext()) {
+            Cache.Entry<String, ValueType> entry = it2.next();
+            assertInstanceOf(ValueType.class, entry.getValue());
+        }
+        Assert.assertFalse("Class Loading Failed", classLoadingFailed);
+    }
+
+    public static class SimulateNonExistantClassLoader extends URLClassLoader {
```

Comment:
This classloader is not used. Is there a test missing?

Reply:
yes, probably.
need to go through history to figure out what happened
the alternative is that there was one and it's no longer applicable, I will take a look

Reply:
well, there is no history since commits were squashed :(

Reply:
I think during the refactoring of tests somehow reference to `SimulateNonExistantClassLoader ` got lost

Reply:
yes, this test is completely broken, I am working on restoring it to functional state

Reply:
ok this is now fixed